File: C++/1215.cpp
__________________________________________________________________________________________________
sample 68 ms submission
class Solution {
public:
  vector<int> countSteppingNumbers(int low, int high) {
    S.clear();
    this->L = high;

    for (int i = 1; i <= 9; ++ i) search(i);

    vector<int> ans;
    if (0 >= low && 0 <= high) {
      ans.push_back(0);
    }
    for (int x : S) {
      if (x >= low && x <= high) ans.push_back(x);
    }
    sort(ans.begin(), ans.end());
    return ans;
  }
private:
  int L;
  vector<int> S;
  void search(long long x) {
    if (x > L) return;
    S.push_back(x);
    int last = x % 10;
    if (last != 0) search(x * 10 + last - 1);
    if (last != 9) search(x * 10 + last + 1);
  }
};
__________________________________________________________________________________________________
sample 76 ms submission
class Solution {
    typedef long long ll;
    vector<ll> s[11];
    void init(int n) {
        if (n == 1)
            for (ll i=0; i<10; i++) s[n].push_back(i);
        else {
            init(n-1);
            for (ll a : s[n-1]) 
                for (ll k : {a%10-1, a%10+1})
                    if (a >= pow(10, n-2) && k>=0 && k<=9) s[n].push_back(a*10+k);
        }
    }
public:
    vector<int> countSteppingNumbers(int low, int high) {
        init(10);
        vector<int> res;
        for (auto& ss : s) 
            for (auto& i : ss)
                if (i>=low && i<=high) res.push_back(i);
        return res;
    }
};
__________________________________________________________________________________________________
sample 140 ms submission
class Solution {
public:
    vector<int> countSteppingNumbers(int low, int high) {
        vector<vector<string>> v(10);
        v[0]=vector<string>({"0","1","2","3","4","5","6","7","8","9"});
        int i, j;
        for(i=1;i<10;i++)
        {
            for(j=0;j<(int)v[i-1].size();j++)
            {
                string s="a"+v[i-1][j];
                if(v[i-1][j][0]!='0')
                {
                    s[0]=v[i-1][j][0]-1;
                    if(i<9||s[0]<'2')
                        v[i].push_back(s);
                }
                if(v[i-1][j][0]!='9')
                {
                    s[0]=v[i-1][j][0]+1;
                    if(i<9||s[0]<'2')
                        v[i].push_back(s);
                }
            }       
        }
        vector<int> r;
        vector<int> result;
        for(i=0;i<10;i++)
            for(j=0;j<(int)v[i].size();j++)
                r.push_back(stoi(v[i][j]));
        sort(r.begin(), r.end());
        for(i=0;i<(int)r.size();i++)
        {
            if(r[i]>=low&&r[i]<=high)
            {
                int x=result.size();
                if(x==0||result[x-1]!=r[i])
                    result.push_back(r[i]);
            }
        }
        return result;
    }
};

File: Java/1215.java
__________________________________________________________________________________________________
sample 16 ms submission
class Solution {
    public int[] countSteppingNumbers(int low, int high) {
        int c1 = ct(low), c2 = ct(high);
        List<Long> re = new ArrayList<>();
        for(int i = c1; i <= c2; i++) {
            get(re, i);
        }
        List<Integer> temp = new ArrayList<>();
        for(long i: re) {
            if(i >= low && i <= high) temp.add((int)(i));
        }
        int[] r = new int[temp.size()];
        for(int i = 0; i < r.length; i++) r[i] = temp.get(i);
        return r;
    }
    int ct(int val) {
        if(val == 0) return 1;
        int re = 0;
        while(val > 0) {
            re++;
            val /= 10;
        }
        return re;
    }
    void get(List<Long> re, int l) {
        if(l == 1) re.add(0l);
        for(long i = 1; i <= 9; i++) dfs(re, 0, i, l);
    }
    void dfs(List<Long> re, int p, long val, int l) {
        if(val < 0) return;
        if(p == l - 1) {
            if(val > Integer.MAX_VALUE) return;
            re.add(val); return;
        }
        if(val % 10 == 0) {
            dfs(re, p + 1, val * 10 + 1, l);
        }
        else if(val % 10 == 9) {
            dfs(re, p + 1, val * 10 + 8, l);
        }
        else {
        dfs(re, p + 1, val * 10 + val % 10 - 1, l);
        dfs(re, p + 1, val * 10 + val % 10 + 1, l);
        }
    }
}
__________________________________________________________________________________________________
22ms
class Solution {
    public List<Integer> countSteppingNumbers(int low, int high) {
        TreeSet<Integer> ans = new TreeSet();
        
        for (int i = 0; i < 10; i++)
            help(i, low, high, ans);
        
        return new ArrayList(ans);
    }
    
    private void help(int n, int low, int high, TreeSet<Integer> ans) {
        if (n <= high && n >= low)
            ans.add(n);
        
        if (n > high || (long) n * 10 >= Integer.MAX_VALUE)
            return;
        
        if (n % 10 != 0)
            help(n % 10 - 1 + n * 10, low, high, ans);
        if (n % 10 != 9)
            help(n % 10 + 1 + n * 10, low, high, ans);
    }
}
__________________________________________________________________________________________________
